src/bytes.ts

```diff
@@ -0,0 +1,265 @@
+}
+
+/**
+ * Assert that a value is a `Uint8Array`.
+ *
+ * @param value - The value to check.
+ * @throws If the value is not a `Uint8Array`.
+ */
+export function assertIsBytes(value: unknown): asserts value is Uint8Array {
+  assert(isBytes(value), 'Value must be a Uint8Array.');
+}
+
+/**
+ * Convert a `Uint8Array` to a hexadecimal string.
+ *
+ * @param bytes - The bytes to convert to a hexadecimal string.
+ * @returns The hexadecimal string.
+ */
+export function bytesToHex(bytes: Uint8Array): string {
+  assertIsBytes(bytes);
+
+  const lookupTable = getPrecomputedHexValues();
+  const hex = Array.prototype.map.call(bytes, (n) => lookupTable[n]).join('');
+
+  return add0x(hex);
+}
+
+/**
+ * Convert a `Uint8Array` to a `bigint`.
+ *
+ * To convert a `Uint8Array` to a `number` instead, use {@link bytesToNumber}.
+ *
+ * @param bytes - The bytes to convert to a bigint.
+ * @returns The bigint.
+ */
+export function bytesToBigInt(bytes: Uint8Array): bigint {
+  assertIsBytes(bytes);
+
+  const hex = bytesToHex(bytes);
+
+  // This can catch possible regressions in the future, if we ever change the
+  // implementation of `bytesToHex` to return a string without the `0x` prefix.
+  assert(hex.startsWith('0x'), 'Hex string must start with "0x".');
+
+  return BigInt(hex);
+}
+
+/**
+ * Convert a `Uint8Array` to a `number`.
+ *
+ * To convert a `Uint8Array` to a `bigint` instead, use {@link bytesToBigInt}.
+ *
+ * @param bytes - The bytes to convert to a number.
+ * @returns The number.
+ * @throws If the resulting number is not a safe integer.
+ */
+export function bytesToNumber(bytes: Uint8Array): number {
+  assertIsBytes(bytes);
+
+  const number = Number(bytesToBigInt(bytes));
+
+  assert(
+    Number.isSafeInteger(number),
+    'Number is not a safe integer. Use `bytesToBigInt` instead.',
+  );
+
+  return number;
+}
+
+/**
+ * Convert a UTF-8 encoded `Uint8Array` to a `string`.
+ *
+ * @param bytes - The bytes to convert to a string.
+ * @returns The string.
+ */
+export function bytesToString(bytes: Uint8Array): string {
+  assertIsBytes(bytes);
+
+  return new TextDecoder(undefined).decode(bytes);
+}
+
+/**
+ * Convert a hexadecimal string to a `Uint8Array`. The string can optionally be
+ * prefixed with `0x`. It accepts even and odd length strings.
+ *
+ * @param value - The hexadecimal string to convert to bytes.
+ * @returns The bytes as `Uint8Array`.
+ */
+export function hexToBytes(value: string): Uint8Array {
+  assertIsHexString(value);
+
+  // Remove the `0x` prefix if it exists, and pad the string to have an even
+  // number of characters.
+  const strippedValue = remove0x(value).toLowerCase();
+  const normalizedValue =
+    strippedValue.length % 2 === 0 ? strippedValue : `0${strippedValue}`;
+  const bytes = new Uint8Array(normalizedValue.length / 2);
+
+  for (let i = 0; i < normalizedValue.length; i++) {
+    // While this is not the prettiest way to convert a hexadecimal string to a
```

Comment:
You might also check if it's a hex string as you go, instead of the above assertion. That'll remove a whole walkthrough of a string

Reply:
You mean something like this?

```ts
assertIsHexString(normalizedValue.slice(i * 2, i * 2 + 1));
```

Reply:
More like
```
const zeroChar = '0'.charCodeAt(0);
const nineChar = '9'.charCodeAt(0);
const aLowChar = 'a'.charCodeAt(0);
const zLowChar = 'z'.charCodeAt(0);
const aHighChar = 'A'.charCodeAt(0);
const zHighChar = 'Z'.charCodeAt(0);
for (let i = 0; i < normalizedValue.length; i++) {
    const c1 = normalizedValue.charCodeAt(i * 2);
    const c2 = normalizedValue.charCodeAt(i * 2 + 1);
    assert((c1 >= zeroChar && c1 <= nineChar) || (c1 >= aLowChar && c1 <= zLowChar) || (c1 >= aHighCar && c1 <= zHighChar));
    assert((c2 >= zeroChar && c1 <= nineChar) || (c2 >= aLowChar && c2 <= zLowChar) || (c2 >= aHighCar && c2 <= zHighChar));
    /* continue as normal */

}
```

Reply:
Just looked more closely at your version. That looks good as well.

Reply:
The downside is that we'd still need to check if `typeof value === 'string'`, or the normalization would fail for certain values. Not sure if it's worth it, I think the performance benefit is minimal.